### PR TITLE
dialects: Improvements to stencil dialect for indexop and applyop

### PR DESCRIPTION
**Pull request by mesham**
Moved dims field from OpAttr IntegerType to IntegerAttr as this was incorrect before. Also can now have zero arguments to the apply op, with the result type provided explicitly now and the number of results too (was previously derived from the first argument)

### CODE REVIEW

**Comment by AntonLydike** [comment]
Could you make the title a bit more concise?

**Comment by mesham** [comment]
> Could you make the title a bit more concise?

Done :)

File: xdsl/dialects/experimental/stencil.py [hunk 1]
```diff
@@ -481,16 +481,13 @@ class ApplyOp(IRDLOperation):
     def get(
         args: Sequence[SSAValue] | Sequence[Operation],
         body: Block,
+        result_type: AnyFloat,
+        result_rank: int,
```
Comment (PapyChacal) [inline xdsl/dialects/experimental/stencil.py:485]: Why not directly a TempType[...] here?

Comment (PapyChacal) [reply xdsl/dialects/experimental/stencil.py:485]: We can have multiple results now too. They would all have the same shape but can hold different element types.

I would go for either making `result_type` a Sequence of types, or completely replace `result_type`, `result_rank` and `result_count` by a `result_types` Sequence of TempType[...].

Note that this is nitpicking, especially for experimental! Happy to merge as-is in the meantime

**Review by PapyChacal — COMMENTED** [review]
Some nitpicking, but happy to merge as-is in experimental! Thanks!
Also, a reminder here that IndexOp is not yet lowered.

**Review by georgebisbas — CHANGES_REQUESTED** [review]
I guess as soon as pyright errors are fixed, it is good to merge

**Review by georgebisbas — APPROVED** [review]

File: xdsl/dialects/experimental/stencil.py (hunk 1 above)
Comment (mesham) [reply xdsl/dialects/experimental/stencil.py]: Thanks, good suggestions :) I have updated the code based on this

**Pull request merged**